Torque frontend for jobcli page (#665)
hpcugent/vsc_user_docs: mkdocs/docs/HPC/torque_frontend_via_jobcli.md

# Torque frontend via jobcli

## What is Torque

> [Torque](https://en.wikipedia.org/wiki/TORQUE) is a resource manager for submitting and managing jobs on an HPC cluster. It is and implementation of [PBS (Portable Batch System)](https://en.wikipedia.org/wiki/Portable_Batch_System). Torque isn't widely used anymore, so the {{hpcinfra}} no longer uses Torque in the backend, however the commands were kept to avoid researchers having to learn other commands.

Reviewer: ```suggestion
[Torque](https://en.wikipedia.org/wiki/TORQUE) is a resource manager for submitting and managing jobs on an HPC cluster. It is an implementation of [PBS (Portable Batch System)](https://en.wikipedia.org/wiki/Portable_Batch_System).
Torque is not widely used anymore, so the {{hpcinfra}} no longer uses Torque in the backend since 2021 in favor of Slurm.
The Torque user interface, which consists of commands like `qsub` and `qstat`, were kept however, to avoid that researchers had to learn other commands to submit and manage jobs.
```

## Slurm backend

> Just like Torque, [Slurm](https://en.wikipedia.org/wiki/Slurm_Workload_Manager) is a resource manager for submitting and managing jobs on an HPC cluster. Currently, Slurm is the most popular workload manager on HPC systems worldwide, however it makes use of different commands that are less intuitive and user friendly than Torque/PBS.

Reviewer: ```suggestion
[Slurm](https://en.wikipedia.org/wiki/Slurm_Workload_Manager) is a resource manager for submitting and managing jobs on an HPC cluster, similar to Torque (but more advanced/modern in some ways). Currently, Slurm is the most popular workload manager on HPC systems worldwide, but it has a user interface that is different and in some sense less user friendly than Torque/PBS.

## jobcli

> Jobcli was developed to make it possible for the {{hpcinfra}} to use a Torque frontend and a Slurm backend. In addition to that, it adds some additional options for Torque commands.

Reviewer: ```suggestion
Jobcli is a Python library that was developed by {{hpcteam}} to make it possible for the {{hpcinfra}} to use a Torque frontend and a Slurm backend. In addition to that, it adds some additional options for Torque commands.
```

Reviewer: This needs a bit more meat: it should also mention that via `jobcli` the commands of Torque user interface (`qsub`, etc. <list them all>) are made available to end users. These correspond to Python script that "translate" the Torque commands to the equivalent Slurm commands, and in the case of `qsub` also make some changes to the provided job script to make it compatible with Slurm (like translating `#PBS` lines to the equivalent `#SBATCH` lines).

> ### help option

Reviewer: This should be under a separate subsection: `### Additional options for Torque commands supported by jobcli`

> Adding `--help` to a Torque command when using it on the {{hpcinfra}} will output a manual for that command, including all possible options for that command (including the original ones from Torque and the ones added by jobcli) and an explanation for each one.

Reviewer: "a manual" -> "an extensive overview of all supported options"

"explanation" -> "short description"

Also, include an example here: show first couple of lines of output produced by `qsub --help`

```
$ qsub --help
...

### dryrun option

> Adding `--dryrun` to a Torque command when using it on the {{hpcinfra}} will show the user what Slurm commands are generated by that Torque command by jobcli. Using `--dryrun` will not actually submit the job.

Reviewer: "submit the job" -> "execute the Slurm backend command"

See [the examples](./#examples)

### debug option

Similarly to `--dryrun`, adding `--debug` to a Torque command when using it on the {{hpcinfra}} will show the user what Slurm commands are generated by that Torque command by jobcli. However in contrast to `--dryrun`, using `--debug` will actually submit the job afterwards.

See [the examples](./#examples)

### Examples

The following examples illustrate the working of the `--dryrun` and `--debug` options with and example jobscript.

`example.sh`:

```shell
#/bin/bash
#PBS -l nodes=1:ppn=8
#PBS -l walltime=2:30:00

> module load SciPy-bundle/...

Reviewer: Use an actual version here, not `...`

```

#### Example of the dryrun option

Running the following command:

```shell
$ qsub --dryrun example.sh -N example
```

will generate this output:

```shell

Command that would have been run:
---------------------------------

/usr/bin/sbatch

Job script that would have been submitted:
------------------------------------------

#!/bin/bash
#SBATCH --chdir="/user/gent/464/vsc46416"
#SBATCH --error="/kyukon/home/gent/464/vsc46416/examples/%x.e%A"
#SBATCH --export="NONE"
#SBATCH --get-user-env="60L"
#SBATCH --job-name="example"
#SBATCH --mail-type="NONE"
#SBATCH --nodes="1"
#SBATCH --ntasks-per-node="8"
#SBATCH --ntasks="8"
#SBATCH --output="/kyukon/home/gent/464/vsc46416/examples/%x.o%A"
#SBATCH --time="02:30:00"

### (start of lines that were added automatically by jobcli)
#
# original submission command:
# qsub --dryrun example.sh -N example
#
# directory where submission command was executed:
# /kyukon/home/gent/464/vsc46416/examples
#
# original script header:
# #PBS -l nodes=1:ppn=8
# #PBS -l walltime=2:30:00
#
### (end of lines that were added automatically by jobcli)

#/bin/bash

module load SciPy-bundle/2023.11-gfbf-2023b

python script.py > script.out.${PBS_JOBID}
```

> This output consist of a few components. For our example the most important lines are the ones that start with `$SBATCH` since these contain the translation of the Torque commands to Slurm commands. For example the job-name is the one we specified with the `-N` option in the command.

Reviewer: You should clearly mention here that the body of the job script is not touched at all: only the header is tweaked, based on the `#PBS` lines in the job script and the options passed to `qsub`.

If a job script uses PBS-isms like `$PBS_JOBID`, they are retained.
Slurm is configured such on the HPC-UGent infrastructure that common `$PBS_*` environment variables are defined in the job environment, next to the Slurm equivalents (this part is not actually done by jobcli currently, but via a Slurm plugin, see also https://github.com/hpcugent/jobcli/issues/19)

# Torque frontend via jobcli

## What is Torque

[Torque](https://en.wikipedia.org/wiki/TORQUE) is a resource manager for submitting and managing jobs on an HPC cluster. It is an implementation of [PBS (Portable Batch System)](https://en.wikipedia.org/wiki/Portable_Batch_System).
Torque is not widely used anymore, so the {{hpcinfra}} no longer uses Torque in the backend since 2021 in favor of Slurm.

> The Torque user interface, which consists of commands like `qsub` and `qstat`, were kept however, to avoid that researchers had to learn other commands to submit and manage jobs.

Reviewer: ```suggestion
The Torque user interface, which consists of commands like `qsub` and `qstat`, was kept however, to avoid that researchers had to learn other commands to submit and manage jobs.

### Additional options for Torque commands supported by jobcli

#### help option

Adding `--help` to a Torque command when using it on the {{hpcinfra}} will output an extensive overview of all supported options for that command, including all possible options for that command (including the original ones from Torque and the ones added by jobcli) and a short description for each one.

For example:
```shell
$ qsub --help
usage: qsub [--version] [--debug] [--dryrun] [--pass OPTIONS] [--dump PATH]...

Submit job script

positional arguments:
  script_file_path      Path to job script to be submitted (default: read job
                        script from stdin)

optional arguments:
  -A ACCOUNT            Charge resources used by this job to specified account
  ...
```

#### dryrun option

Adding `--dryrun` to a Torque command when using it on the {{hpcinfra}} will show the user what Slurm commands are generated by that Torque command by jobcli. Using `--dryrun` will not actually execute the Slurm backend command.

> See [the examples](./#examples)

Reviewer: ```suggestion
See also [the examples](./#examples) below.

#### debug option

> Similarly to `--dryrun`, adding `--debug` to a Torque command when using it on the {{hpcinfra}} will show the user what Slurm commands are generated by that Torque command by jobcli. However in contrast to `--dryrun`, using `--debug` will actually submit the job afterwards.

Reviewer: ```suggestion
Similarly to `--dryrun`, adding `--debug` to a Torque command when using it on the {{hpcinfra}} will show the user what Slurm commands are generated by that Torque command by jobcli. However in contrast to `--dryrun`, using `--debug` will actually run the Slurm backend command.

> See [the examples](./#examples)

Reviewer: ```suggestion
See also [the examples](./#examples) below.

#### Examples

> The following examples illustrate the working of the `--dryrun` and `--debug` options with and example jobscript.

Reviewer: The following examples illustrate the working of the `--dryrun` and `--debug` options with an example jobscript.

`example.sh`:

```shell
#/bin/bash
#PBS -l nodes=1:ppn=8
#PBS -l walltime=2:30:00

module load SciPy-bundle/2023.11-gfbf-2023b

python script.py > script.out.${PBS_JOBID}
```

##### Example of the dryrun option

Running the following command:

```shell
$ qsub --dryrun example.sh -N example
```

will generate this output:

```shell

Command that would have been run:
---------------------------------

/usr/bin/sbatch

Job script that would have been submitted:
------------------------------------------

#!/bin/bash
#SBATCH --chdir="/user/gent/464/vsc46416"

> #SBATCH --error="/kyukon/home/gent/464/vsc46416/examples/%x.e%A"

Author: Should I remove my account name from these examples?

#SBATCH --export="NONE"
#SBATCH --get-user-env="60L"
#SBATCH --job-name="example"
#SBATCH --mail-type="NONE"
#SBATCH --nodes="1"
#SBATCH --ntasks-per-node="8"
#SBATCH --ntasks="8"
#SBATCH --output="/kyukon/home/gent/400/{{userid}}/examples/%x.o%A"
#SBATCH --time="02:30:00"

### (start of lines that were added automatically by jobcli)
#
# original submission command:
# qsub --dryrun example.sh -N example
#
# directory where submission command was executed:
# /kyukon/home/gent/400/{{userid}}/examples
#
# original script header:
# #PBS -l nodes=1:ppn=8
# #PBS -l walltime=2:30:00
#
### (end of lines that were added automatically by jobcli)

#/bin/bash

module load SciPy-bundle/2023.11-gfbf-2023b

python script.py > script.out.${PBS_JOBID}
```

> This output consist of a few components. For our example the most important lines are the ones that start with `$SBATCH` since these contain the translation of the Torque commands to Slurm commands. For example the job-name is the one we specified with the `-N` option in the command.

Reviewer: ```suggestion
This output consist of a few components. For our example the most important lines are the ones that start with `#SBATCH` since these contain the translation of the Torque commands to Slurm commands. For example the job-name is the one we specified with the `-N` option in the command.
